### PR TITLE
os/se/ameba: SE Support GCM mode

### PR DESCRIPTION
**Pull request by ZhenBei-Sin**
Change Notes:
- Added GCM mode to support GCM encryption/decryption in SE API
- Added smccc entry for the GCM algorithm

### CODE REVIEW

**Comment by ZhenBei-Sin** [comment]
Hi @sunghan-chang  & @jylee9613 

The GCM mode was ready. Please help to verify.
For the TP1x: 
PR was submitted for the SE lib with included the GCM mode.
Please re-gen the KM4 img3 from the PR and commit to TizenRT.

Thank You.

Best Regards,
Zhen Bei

**Comment by ZhenBei-Sin** [comment]
> @ZhenBei-Sin How about separating the pr for `implement secure reboot reason`?

The changes was affected the secure BL image. So put under same PR with separate commit.

**Comment by ZhenBei-Sin** [comment]
> > @ZhenBei-Sin How about separating the pr for `implement secure reboot reason`?
> 
> FYI: The changes was affected the secure BL image. So put under same PR with separate commit. If require to separate PR, it will do it…
> 
> Thank you.

Hi @sunghan-chang 
Could you let me know, still want to remain or separate PR?
Thank You.

**Comment by ZhenBei-Sin** [comment]
> @ZhenBei-Sin GCM could be merged atleast 1 and half week later. How about merging the reboot reason first?

Hi @sunghan-chang 
The reboot reason is for the secure part and the implementation is secure.
I will proceed to separate the PR.
1. With implementation secure reboot reason.
2. GCM. (will include the 1. changes in SE)

Any concern ?

**Comment by sunghan-chang** [comment]
> > @ZhenBei-Sin GCM could be merged atleast 1 and half week later. How about merging the reboot reason first?
> 
> Hi @sunghan-chang The reboot reason is for the secure part and the implementation is secure. I will proceed to separate the PR.
> 
> 1. With implementation secure reboot reason.
> 2. GCM. (will include the 1. changes in SE)
> 
> Any concern ?

It looks ok.

**Comment by ZhenBei-Sin** [comment]
> > > @ZhenBei-Sin GCM could be merged atleast 1 and half week later. How about merging the reboot reason first?
> > 
> > 
> > Hi @sunghan-chang The reboot reason is for the secure part and the implementation is secure. I will proceed to separate the PR.
> > 
> > 1. With implementation secure reboot reason.
> > 2. GCM. (will include the 1. changes in SE)
> > 
> > Any concern ?
> 
> It looks ok.

Done separate the PR. PR#6274.

**Comment by sunghan-chang** [comment]
NOTE!! This must be merged after PR #6274 .

File: os/se/ameba/security_ameba_wrapper_tz.c
Comment (ZhenBei-Sin) [reply os/se/ameba/security_ameba_wrapper_tz.c:745]: Hi @jylee9613 
May I know u using which config ? flat_apps or loadable_apps or loadable_ext?

Tmr, I will test with ur given enable config.

Thank you.

Comment (ZhenBei-Sin) [reply os/se/ameba/security_ameba_wrapper_tz.c:745]: Hi @jylee9613 

I tested the latest TizenRT commit without AES GCM, and it was able to reproduce the issue with the given config.

`CONFIG_DEBUG_WARN=y
CONFIG_DEBUG_VERBOSE=y
CONFIG_DEBUG_SECURE_ELEMENT_ERROR=y
CONFIG_DEBUG_SECURE_ELEMENT_INFO=y
CONFIG_HW_AES_ENC=y
CONFIG_HW_GCM_ENC=y`

So, it can be concluded that the issue is not related to the AES GCM implementation.
Upon investigation, the issue occurs when both CONFIG_DEBUG_WARN=y and CONFIG_DEBUG_VERBOSE=y are enabled.
If either of these configs is enabled alone, the issue does not reproduce.

Thank You.

Comment (jylee9613) [reply os/se/ameba/security_ameba_wrapper_tz.c:745]: Hello @ZhenBei-Sin 
Thank you for checking. Then I'll test with enabling one debug option. 

Thank you:)

Comment (ZhenBei-Sin) [reply os/se/ameba/security_ameba_wrapper_tz.c:745]: Hi @jylee9613 ,

This is the new findings:
/os/include/debug.h
os/drivers/seclink/seclink_drv.c

Remove the "vdbg" from the "se_register" with using the config that enables all options you provided. 
It is unable to reproduce the issue.

`
int se_register(const char *path, struct sec_lowerhalf_s *lower)
{
	//vdbg("Registering %s\n", path);

	if (!lower) {
		return -1;
	}
`

Therefore, it can be concluded that the issue is related to the debug print statement. 
Please investigate further on this matter.

Thank You.

Comment (edwakuwaku) [reply os/se/ameba/security_ameba_wrapper_tz.c:745]: Hi @jylee9613 , as we knew, `lldbg` is used for booting sequence log. How about try to change this to use "lldbg" instead of "vdbg"? Maybe it is able to resolve hang issue?

**Comment by jylee9613** [comment]
I'm writing test codes to test GCM function, but I got some errors while testing, so it takes some time. I'm sorry for late update, and I'll let you know the test result asap. Thank you.

**Comment by ZhenBei-Sin** [comment]
This is need to re-gen and re-base after the DDR PR.

**Comment by ZhenBei-Sin** [comment]
> This is need to re-gen and re-base after the DDR PR.



> This is need to re-gen and re-base after the DDR PR.

NOTE:
Done rebase and solved the conflict.
LGTM to merge once done verification of GCM mode from Samsung side.
Thank you.

**Review by jylee9613 — APPROVED** [review]
I confirmed that all operation works properly in TP1x and AIDual.
Thank you for support and waiting for us.
After merge this PR, I'll merge PRs in Trustzone too.

**Review by sunghan-chang — APPROVED** [review]

**Comment by sunghan-chang** [comment]
@ZhenBei-Sin This PR looks Aug 16th code base. Could you finally confirm whether no need to rebase latest again or not?
I mean,
can I merge this now?

**Comment by ZhenBei-Sin** [comment]
> @ZhenBei-Sin This PR looks Aug 16th code base. Could you finally confirm whether no need to rebase latest again or not? I mean, can I merge this now?

Hi @sunghan-chang ,

Please refer to the [selib_releasenote.txt] in the ecode.
From August 16th until now, does not contain any changes related to SE, so it looks good to merge.
Thank You.

**Pull request merged**
